torch/csrc/jit/codegen/cuda/iter_visitor.cpp

```diff
@@ -52,46 +84,47 @@ void remove_visited(
 void IterVisitor::traverseFrom(
     Fusion* const fusion,
     const std::vector<Val*>& from,
-    bool traverseAllPaths) {
+    bool traverseAllPaths,
+    bool respectComputeAt) {
   FusionGuard fg(fusion);
   std::unordered_set<Statement*> visited;
   stmt_stack.clear();
-  if (!from.empty())
+  if (!from.empty()) {
     stmt_stack.emplace_back(from.rbegin(), from.rend());
+  }
+  // true when returning to a node after vistiting all its input
+  // nodes. Nodes are only visited when this is true.
+  bool all_inputs_visited = false;
 
   while (!stmt_stack.empty()) {
-    auto next_stmts = next(stmt_stack.back().back());
-
-    // Remove statements we already visited if we're not traversing all paths
-    if (!traverseAllPaths)
-      remove_visited(next_stmts, visited);
-
-    // Traverse down until we get to a leaf
-    while (!next_stmts.empty()) {
-      stmt_stack.emplace_back(next_stmts.rbegin(), next_stmts.rend());
-      next_stmts = next(stmt_stack.back().back());
-      // Remove statements we already visited if we're not traversing all paths
+    auto& current_inputs = stmt_stack.back();
+    // When current_inputs is empty, all the input nodes have been
+    // visited. Return to the output node by popping the stack. Record
+    // all inputs are visited.
+    if (current_inputs.empty()) {
+      stmt_stack.pop_back();
+      all_inputs_visited = true;
+      continue;
+    }
+    auto& stmt = current_inputs.back();
+    // Visit stmt when all_inputs_visited is true.
+    if (all_inputs_visited) {
+      // Mark visited
+      visited.emplace(stmt);
```

Comment:
I'd use plain insert() here - we're inserting just a pointer, right? (emplace() suggests that there's an expensive construction going on)